src/btree/bt_discard.c

```diff
@@ -531,3 +531,30 @@ __wt_free_update_list(WT_SESSION_IMPL *session, WT_UPDATE **updp)
     }
     *updp = NULL;
 }
+
+/*
+ * __wt_free_obsolete_updates --
+ *     Following a globally visible update, free any obsolete updates in the update chain. After a
+ *     globally visible update, no reader finds any updates. It is the responsibility of the caller
+ *     to lock the page before freeing the updates.
```

Comment:
I don't think the `page_lock` is adequate to make this safe. The `page_lock` is only acquired by operations that are making changes to the page - inserting new content, splitting a page and reconciliation are the three things that acquire the page lock.

Readers in the system don't acquire the page lock - so don't get any protection from it being held. i.e: a reader can traverse into an update chain without acquiring the page lock, and I think might have traversed to an update that has since been marked as obsolete (I'm happy to be told that I'm wrong..)

Reply:
It might not be related to the change here specifically, but do we need a mechanism that ensures no reader traverses an update chain without transaction visibility setup in it's local transaction structure? That feels like an implicit contract we could get wrong at the moment (or maybe it isn't implicit?)

Reply:
Apologies for the late reply.
We do not allow a transaction to traverse the update chain without a snapshot, except when it is read-uncommitted isolation. I agree that if we allow a transaction to traverse the update chain without visibility, information can go wrong, but currently it is not possible, so there is no problem.